README.md
![bitmapist](https://raw.githubusercontent.com/Doist/bitmapist4/master/static/bitmapist.png "bitmapist")


[![Build Status](https://travis-ci.org/Doist/bitmapist4.svg?branch=master)](https://travis-ci.org/Doist/bitmapist4)

**NEW!** Try out our new standalone [bitmapist-server](https://github.com/Doist/bitmapist-server), which improves memory efficiency 443 times and makes your setup cheaper and more scaleable. It's fully compatable with bitmapist that runs on Redis.

# bitmapist: a powerful analytics library for Redis

This Python library makes it possible to implement real-time, highly scalable analytics that can answer following questions:

* Has user 123 been online today? This week? This month?
* Has user 123 performed action "X"?
* How many users have been active have this month? This hour?
* How many unique users have performed action "X" this week?
* How many % of users that were active last week are still active?
* How many % of users that were active last month are still active this month?
* What users performed action "X"?

This library is very easy to use and enables you to create your own reports easily.

Using Redis bitmaps you can store events for millions of users in a very little amount of memory (megabytes).
You should be careful about using huge ids as this could require larger amounts of memory. Ids should be in range [0, 2^32).

Additionally bitmapist can generate cohort graphs that can do following:
* Cohort over user retention
* How many % of users that were active last [days, weeks, months] are still active?
* How many % of users that performed action X also performed action Y (and this over time)
* And a lot of other things!

If you want to read more about bitmaps please read following:

* http://blog.getspool.com/2011/11/29/fast-easy-realtime-metrics-using-redis-bitmaps/
* http://redis.io/commands/setbit
* http://en.wikipedia.org/wiki/Bit_array
* http://www.slideshare.net/crashlytics/crashlytics-on-redis-analytics



# Installation

Can be installed very easily via:

    $ pip install bitmapist4


# Ports

* PHP port: https://github.com/jeremyFreeAgent/Bitter


# Examples

Setting things up:

```python
import bitmapist4
b = bitmapist4.Bitmapist()
```

Mark user 123 as active and has played a song:

```python
b.mark_event('active', 123)
b.mark_event('song:played', 123)
```

Answer if user 123 has been active this month:

```python
assert 123 in b.MonthEvents('active')
assert 123 in b.MonthEvents('song:played')
```


How many users have been active this week?:

```python
len(b.WeekEvents('active'))
```

Iterate over all users active this week:

```python
for uid in b.WeekEvents('active'):
    print(uid)
```


To explore any specific day, week, month or year instead of the current one, 
uou can create an event from any datetime object with a `from_date` static
method.

```python
specific_date = datetime.datetime(2018, 1, 1)
ev = b.MonthEvents('active').from_date(specific_date)
print(len(ev))
```

There are methods `prev` and `next` returning "sibling" events and
allowing you to walk through events in time without any sophisticated
iterators. A `delta` method allows you to jump forward or backward for
more than one step. Uniform API allows you to use all types of base events
(from hour to year) with the same code.

```python

current_month = b.MonthEvents('active')
prev_month = current_month.prev()
next_month = current_month.next()
year_ago = current_month.delta(-12)
```

Every event object has `period_start` and `period_end` methods to find a
time span of the event. This can be useful for caching values when the caching
of "events in future" is not desirable:

```python

ev = b.MonthEvent('active', dt)
if ev.period_end() < datetime.datetime.utcnow():
    cache.set('active_users_<...>', len(ev))
```


Tracking hourly is disabled (to save memory!) You can enable it with a
constructor argument.

```python
b = bitmapist4.Bitmapist(track_hourly=True)
```

Additionally you can supply an extra argument to `mark_event` to bypass the default value::

```python
b.mark_event('active', 123, track_hourly=False)
```


## Unique events

Sometimes data of the event makes little or no sense and you are more interested
if that specific event happened at least once in a lifetime for a user. 

There is a `UniqueEvents` model for this purpose. The model creates only one
Redis key and doesn't depend on the date.

You can combine unique events with other types of events.

A/B testing example:

```python

active = b.DailyEvents('active')
a = b.UniqueEvents('signup_form:classic')
b = b.UniqueEvents('signup_form:new')

print("Active users, signed up with classic form", len(active & a))
print("Active users, signed up with new form", len(active & b))
```

You can mark these users with `b.mark_unique` or you can automatically
populate the extra unique cohort for all marked keys

```python
b = bitmapist4.Bitmapist(track_unique=True)
b.mark_event('premium', 1)
assert 1 in b.UniqueEvents('premium')
``` 

## Perform bit operations

How many users that have been active last month are still active this month?

```python
ev = b.MonthEvents('active')
active_2months = ev & ev.prev() 
print(len(active_2months))

# Is 123 active for 2 months?
assert 123 in active_2months
```

Operators `&`, `|`, `^` and `~` supported.

This works with nested bit operations (imagine what you can do with this ;-))!


## Delete events

If you want to permanently remove marked events for any time period you can use the `delete()` method:

```python
ev = b.MonthEvents.from_date('active', last_month)
ev.delete()
```

If you want to remove all bitmapist events use:
```python
b.delete_all_events()
```

Results of bit operations are cached by default. They're cached for 60 seconds
for operations, contained non-finished periods, and for 24 hours otherwise.

You may want to reset the cache explicitly:

```python
ev = b.MonthEvents('active')
active_2months = ev & ev.prev() 
# Delete the temporary AND operation
active_2months.delete()

# delete all bit operations (slow if you have many millions of keys in Redis)
b.delete_temporary_bitop_keys()
```

## Bulk updates with transactions

If you often performs multiple updates at once, you can benefit from Redis
pipelines, wrapped as transactions inside bitmapist.

```python
with b.transaction():
    b.mark_event('active')
    b.mark_event('song:played')
```


# Migration from previous version

The API of the "bitmapist4.Bitmapist" instance is mostly compatible with the
API of previous version of bitmapist (module-level). Notable changes outlined
below.

- Removed the "system" attribute for choosing the server. You are supposed to
  use different Bitmapist class instances instead. If you used "system" to
  work with pipelines, you should switch to transactions instead.
- bitmapist.TRACK_HOURLY and bitmapist.TRACK_UNIQUE module-level constants
  moved to bitmapist4.Bitmapist attributes and can be set up with a class
  constructor.
- On a database level, new bitmapist4 uses "bitmapist_" prefix for Redis keys,
  while old bitmapist uses "trackist_" for historical reasons. If you want
  to keep using the old database, or want to use bitmapist and bitmapist4
  against the same database, you need to explicitly set the key prefix
  to "trackist_".
- If you use bitmapist-server, make sure that you use the version 1.2 or newer.
  This version adds the support for  EXPIRE command which is used to expire
  temporary bitop keys.


Replace old code which could look like this:

```python
import bitmapist
bitmapist.setup_redis('default', 'localhost', 6380)
...
bitmapist.mark_event('acive', user_id)
```

With something looking like this:

```python
from bitmapist4 import Bitmapist
bitmapist = Bitmapist('redis://localhost:6380', key_prefix='trackist_')
...
bitmapist.mark_event('acive', user_id)
```



# Bitmapist cohort

Cohort is a group of subjects who share a defining characteristic (typically
subjects who experienced a common event in a selected time period, such as
birth or graduation).

You can get the cohort table using `bitmapist4.cohort.get_cohort_table()` 
function. 

Each row of this table answers the question "what part of the `cohort`
performed `activity` over time", and Nth cell of that row represents the
number of users (absolute or in percent) which still perform the activity
N days (or weeks, or months) after.

Each new column of the cohort unfolds the behavior of different similar
cohorts over time. The latest row displays the behavior of the cohort,
provided as an argument, the one above displays the behavior of the similar
cohort, but shifted 1 day (or week, or month) ago, etc.

For example, consider following cohort statistics

```
table = get_cohort_table(b.WeekEvents('registered'), b.WeekEvents('active'))
```

This table shows what's the rate of registered users is still active
the same week after registration, then one week after, then two weeks
after the registration, etc.

By default the table displays 20 rows.

The first row represents the statistics from cohort of users, registered
20 weeks ago. The second row represents the same statistics for users,
registered 19 week ago, and so on until finally the latest row shows users
registered this week. Naturally, the last row will contain only one cell,
the number of users that were registered this week AND were active this
week as well.


Then you may render it yourself to HTML, or export to Pandas dataframe
with df() method.

Sample from user activity on http://www.gharchive.org/

```python
In [1]: from bitmapist4 import Bitmapist, cohort

In [2]: b = Bitmapist()

In [3]: cohort.get_cohort_table(b.WeekEvents('active'), b.WeekEvents('active'), rows=5, use_percent=False).df()
Out[3]:
             cohort       0        1        2        3        4
05 Nov 2018  137420  137420  25480.0  18358.0  21575.0  18430.0
12 Nov 2018  150975  150975  22195.0  25833.0  21165.0      NaN
19 Nov 2018  121417  121417  22477.0  15796.0      NaN      NaN
26 Nov 2018  152027  152027  25606.0      NaN      NaN      NaN
03 Dec 2018  130470  130470      NaN      NaN      NaN      NaN
```

The dataframe can be further colorized (to be displayed in Jupyter notebooks)
with stylize().


---

Copyright: 2012-2018 by Doist Ltd.

License: BSD


CHANGELOG
## [4.0] - 2018-12-11

Released the first beta version of bitmapist4. Main differences from the
previous version.

- Class-based configuration. To work with a specific storage and to set
  some specific settings, a Bitmapist object is instantiated
  instead of using global SYSTEM

- Support for automatic and explicit expiration for temporary keys
  generated on bit operations, and especially during cohort calculation

- Cohort package rewritten to make more convenient usage with wider range of
  scenarios

- Add support for transactions (wrappers around Redis pipelines)

- Tests run against Redis and Bitmapist-server


test/conftest.py
from builtins import range

import pytest
import os
import subprocess
import atexit
import socket
import time
import redis
import bitmapist4

REDIS_HOST = '127.0.0.1'
REDIS_PORT = 6399

BITMAPIST_SERVER_HOST = '127.0.0.1'
BITMAPIST_SERVER_PORT = 6400


@pytest.fixture(scope='session', autouse=True, params=['redis', 'bitmapist-server'])
def redis_server(request):
    """
    Fixture starting Redis or bitmapist-server process
    """
    if request.param == 'redis':
        host, port = REDIS_HOST, REDIS_PORT
    else:
        host, port = BITMAPIST_SERVER_HOST, BITMAPIST_SERVER_PORT

    if is_socket_open(host, port):
        yield host, port
    else:
        proc = start_server(request.param, host, port)
        wait_for_socket(host, port)
        yield host, port
        proc.terminate()


@pytest.fixture
def bitmapist(redis_server):
    conn = redis.StrictRedis(*redis_server)
    obj = bitmapist4.Bitmapist(conn, track_hourly=True)
    yield obj
    flushall(conn)


@pytest.fixture
def bitmapist_non_unique(redis_server):
    conn = redis.StrictRedis(*redis_server)
    obj = bitmapist4.Bitmapist(conn, track_hourly=True, track_unique=False)
    yield obj
    flushall(conn)


@pytest.fixture
def bitmapist_copy(redis_server):
    conn = redis.StrictRedis(*redis_server)
    obj = bitmapist4.Bitmapist(conn)
    yield obj
    flushall(conn)


@pytest.fixture
def db1(redis_server):
    host, port = redis_server
    conn = redis.StrictRedis(host, port, db=1)
    obj = bitmapist4.Bitmapist(conn)
    yield obj
    flushall(conn)


def flushall(conn):
    """
    bitmapist-server-compatible command to delete all keys from the server
    """
    keys = conn.keys('*')
    if keys:
        conn.delete(*keys)


def start_server(server_type, host, port):
    if server_type == 'redis':
        return start_redis_server(host, port)
    else:
        return start_bitmapist_server(host, port)


def start_redis_server(host, port):
    """
    Helper function starting Redis server
    """
    devzero = open(os.devnull, 'r')
    devnull = open(os.devnull, 'w')
    proc = subprocess.Popen(
        ['redis-server', '--bind', host, '--port', str(port)],
        stdin=devzero,
        stdout=devnull,
        stderr=devnull,
        close_fds=True)
    atexit.register(lambda: proc.terminate())
    return proc


def start_bitmapist_server(host, port):
    """
    Helper function starting bitmapist server
    """
    devzero = open(os.devnull, 'r')
    devnull = open(os.devnull, 'w')
    proc = subprocess.Popen(
        ['bitmapist-server', '-addr', '{}:{}'.format(host, port)],
        stdin=devzero,
        stdout=devnull,
        stderr=devnull,
        close_fds=True)
    atexit.register(lambda: proc.terminate())
    return proc


def is_socket_open(host, porto):
    """
    Helper function which tests is the socket open
    """
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.settimeout(0.1)
    ok = sock.connect_ex((host, porto)) == 0
    sock.close()
    return ok


def wait_for_socket(host, porto, seconds=3):
    """
    Check if socket is up for :param:`seconds` sec, raise an error otherwise
    """
    polling_interval = 0.2
    iterations = int(seconds / polling_interval)

    time.sleep(polling_interval)
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.settimeout(0.1)
    for _ in range(iterations):
        time.sleep(polling_interval)
        result = sock.connect_ex((host, porto))
        if result == 0:
            sock.close()
            break
    else:
        raise RuntimeError(
            'Service at %s:%d is unreachable' % (host, porto))


test/test_equality.py
def test_equality(bitmapist):
    ev1 = bitmapist.YearEvents('foo', 2014)
    ev2 = bitmapist.YearEvents('foo', 2014)
    ev3 = bitmapist.YearEvents('foo', 2015)
    assert ev1 == ev2
    assert ev1 != ev3


test/test_bitmapist.py
from datetime import datetime, timedelta


def test_mark_with_diff_days(bitmapist):
    bitmapist.mark_event('active', 123)

    # Month
    assert 123 in bitmapist.MonthEvents.from_date('active')
    assert 124 not in bitmapist.MonthEvents.from_date('active')

    # Week
    assert 123 in bitmapist.WeekEvents.from_date('active')
    assert 124 not in bitmapist.WeekEvents.from_date('active')

    # Day
    assert 123 in bitmapist.DayEvents.from_date('active')
    assert 124 not in bitmapist.DayEvents.from_date('active')

    # Hour
    assert 123 in bitmapist.HourEvents.from_date('active')
    assert 124 not in bitmapist.HourEvents.from_date('active')
    assert 124 not in bitmapist.HourEvents.from_date('active').prev()


def test_mark_unmark(bitmapist):
    now = datetime.utcnow()

    bitmapist.mark_event('active', 125)
    assert 125 in bitmapist.MonthEvents('active', now.year, now.month)

    bitmapist.unmark_event('active', 125)
    assert 125 not in bitmapist.MonthEvents('active', now.year, now.month)


def test_mark_counts(bitmapist):
    now = datetime.utcnow()

    assert bitmapist.MonthEvents('active', now.year,
                                 now.month).get_count() == 0

    bitmapist.mark_event('active', 123)
    bitmapist.mark_event('active', 23232)

    assert len(bitmapist.MonthEvents('active', now.year, now.month)) == 2


def test_mark_iter(bitmapist):
    now = datetime.utcnow()
    ev = bitmapist.MonthEvents('active', now.year, now.month)

    assert list(ev) == []

    bitmapist.mark_event('active', 5)
    bitmapist.mark_event('active', 55)
    bitmapist.mark_event('active', 555)
    bitmapist.mark_event('active', 5555)

    assert list(ev) == [5, 55, 555, 5555]


def test_different_dates(bitmapist):
    now = datetime.utcnow()
    yesterday = now - timedelta(days=1)

    bitmapist.mark_event('active', 123, timestamp=now)
    bitmapist.mark_event('active', 23232, timestamp=yesterday)

    assert bitmapist.DayEvents('active', now.year, now.month,
                               now.day).get_count() == 1

    assert bitmapist.DayEvents('active', yesterday.year, yesterday.month,
                               yesterday.day).get_count() == 1


def test_different_buckets(bitmapist):
    now = datetime.utcnow()

    bitmapist.mark_event('active', 123)
    bitmapist.mark_event('tasks:completed', 23232)

    assert bitmapist.MonthEvents('active', now.year,
                                 now.month).get_count() == 1
    assert bitmapist.MonthEvents('tasks:completed', now.year,
                                 now.month).get_count() == 1


def test_bit_operations(bitmapist, bitmapist_copy):
    now = datetime.utcnow()
    last_month = datetime.utcnow() - timedelta(days=30)

    # 123 has been active for two months
    bitmapist.mark_event('active', 123, timestamp=now)
    bitmapist.mark_event('active', 123, timestamp=last_month)

    # 224 has only been active last_month
    bitmapist.mark_event('active', 224, timestamp=last_month)

    # Assert basic premises
    assert bitmapist.MonthEvents('active', last_month.year,
                                 last_month.month).get_count() == 2
    assert bitmapist.MonthEvents('active', now.year,
                                 now.month).get_count() == 1

    # Try out with bit AND operation
    active_2_months = bitmapist_copy.BitOpAnd(
        bitmapist.MonthEvents('active', last_month.year, last_month.month),
        bitmapist.MonthEvents('active', now.year, now.month))
    assert active_2_months.get_count() == 1
    assert 123 in active_2_months
    assert 224 not in active_2_months
    active_2_months.delete()

    # Try out with bit OR operation
    assert bitmapist.BitOpOr(
        bitmapist.MonthEvents('active', last_month.year, last_month.month),
        bitmapist.MonthEvents('active', now.year, now.month)).get_count() == 2

    # Try out with a different system
    active_2_months = bitmapist.BitOpAnd(
        bitmapist.MonthEvents('active', last_month.year, last_month.month),
        bitmapist.MonthEvents('active', now.year, now.month),
    )
    assert active_2_months.get_count() == 1
    active_2_months.delete()

    # Try nested operations
    active_2_months = bitmapist.BitOpAnd(
        bitmapist.BitOpAnd(
            bitmapist.MonthEvents('active', last_month.year, last_month.month),
            bitmapist.MonthEvents('active', now.year, now.month)),
        bitmapist.MonthEvents('active', now.year, now.month))

    assert 123 in active_2_months
    assert 224 not in active_2_months
    active_2_months.delete()


def test_bit_operations_complex(bitmapist):
    now = datetime.utcnow()
    tom = now + timedelta(days=1)

    bitmapist.mark_event('task1', 111, timestamp=now)
    bitmapist.mark_event('task1', 111, timestamp=tom)
    bitmapist.mark_event('task2', 111, timestamp=now)
    bitmapist.mark_event('task2', 111, timestamp=tom)
    bitmapist.mark_event('task1', 222, timestamp=now)
    bitmapist.mark_event('task1', 222, timestamp=tom)
    bitmapist.mark_event('task2', 222, timestamp=now)
    bitmapist.mark_event('task2', 222, timestamp=tom)

    now_events = bitmapist.BitOpAnd(
        bitmapist.DayEvents('task1', now.year, now.month, now.day),
        bitmapist.DayEvents('task2', now.year, now.month, now.day))

    tom_events = bitmapist.BitOpAnd(
        bitmapist.DayEvents('task1', tom.year, tom.month, tom.day),
        bitmapist.DayEvents('task2', tom.year, tom.month, tom.day))

    both_events = bitmapist.BitOpAnd(now_events, tom_events)

    assert len(now_events) == len(tom_events)
    assert len(now_events) == len(both_events)


def test_bitop_key_sharing(bitmapist):
    today = datetime.utcnow()

    bitmapist.mark_event('task1', 111, timestamp=today)
    bitmapist.mark_event('task2', 111, timestamp=today)
    bitmapist.mark_event('task1', 222, timestamp=today)
    bitmapist.mark_event('task2', 222, timestamp=today)

    ev1_task1 = bitmapist.DayEvents('task1', today.year, today.month,
                                    today.day)
    ev1_task2 = bitmapist.DayEvents('task2', today.year, today.month,
                                    today.day)
    ev1_both = bitmapist.BitOpAnd(ev1_task1, ev1_task2)

    ev2_task1 = bitmapist.DayEvents('task1', today.year, today.month,
                                    today.day)
    ev2_task2 = bitmapist.DayEvents('task2', today.year, today.month,
                                    today.day)
    ev2_both = bitmapist.BitOpAnd(ev2_task1, ev2_task2)

    assert ev1_both.redis_key == ev2_both.redis_key
    assert len(ev1_both) == len(ev1_both) == 2
    ev1_both.delete()
    assert len(ev1_both) == len(ev1_both) == 0


def test_events_marked(bitmapist):
    now = datetime.utcnow()

    assert bitmapist.MonthEvents('active', now.year,
                                 now.month).get_count() == 0
    assert not bitmapist.MonthEvents('active', now.year,
                                     now.month).has_events_marked()

    bitmapist.mark_event('active', 123, timestamp=now)

    assert bitmapist.MonthEvents('active', now.year,
                                 now.month).get_count() == 1
    assert bitmapist.MonthEvents('active', now.year,
                                 now.month).has_events_marked()


def test_get_event_names(bitmapist):
    event_names = {'foo', 'bar', 'baz', 'spam', 'egg'}
    for e in event_names:
        bitmapist.mark_event(e, 1)
    bitmapist.BitOpAnd(bitmapist.DayEvents('foo'), bitmapist.DayEvents('bar'))
    assert set(bitmapist.get_event_names(batch=2)) == event_names


def test_get_event_names_prefix(bitmapist):
    event_names = {'foo', 'bar', 'baz', 'spam', 'egg'}
    for e in event_names:
        bitmapist.mark_event(e, 1)
    bitmapist.BitOpAnd(bitmapist.DayEvents('foo'), bitmapist.DayEvents('bar'))
    assert set(bitmapist.get_event_names(prefix='b',
                                         batch=2)) == {'bar', 'baz'}


def test_bit_operations_magic(bitmapist):
    bitmapist.mark_event('foo', 1)
    bitmapist.mark_event('foo', 2)
    bitmapist.mark_event('bar', 2)
    bitmapist.mark_event('bar', 3)
    foo = bitmapist.DayEvents('foo')
    bar = bitmapist.DayEvents('bar')
    assert list(foo & bar) == [2]
    assert list(foo | bar) == [1, 2, 3]
    assert list(foo ^ bar) == [1, 3]
    assert list(~foo & bar) == [3]


def test_year_events(bitmapist):
    bitmapist.mark_event('foo', 1)
    assert 1 in bitmapist.YearEvents('foo')


bitmapist4/core.py
# -*- coding: utf-8 -*-
from contextlib import contextmanager
try:
    from typing import Type
except ImportError:  # Python 2.x
    pass

import redis
import datetime
from bitmapist4 import events as ev


class Bitmapist(object):
    """
    Core bitmapist object
    """

    # Should hourly be tracked as default?
    # Note that this can have huge implications in amounts
    # of memory that Redis uses (especially with huge integers)
    track_hourly = False

    # Should unique events be tracked as default?
    track_unique = False

    def __init__(self,
                 connection_or_url=redis.StrictRedis(),
                 track_hourly=False,
                 track_unique=True,
                 finished_ops_expire=3600 * 24,
                 unfinished_ops_expire=60,
                 key_prefix='bitmapist_'):
        if isinstance(connection_or_url, redis.StrictRedis):
            self.connection = connection_or_url
        else:
            self.connection = redis.StrictRedis.from_url(connection_or_url)
        self.track_hourly = track_hourly
        self.track_unique = track_unique
        self.finished_ops_expire = finished_ops_expire
        self.unfinished_ops_expire = unfinished_ops_expire
        self.key_prefix = key_prefix
        self.pipe = None

        kw = {'bitmapist': self}
        self.UniqueEvents = type('UniqueEvents', (ev.UniqueEvents, ),
                                 kw)  # type: Type[ev.UniqueEvents]
        self.YearEvents = type('YearEvents', (ev.YearEvents, ),
                               kw)  # type: Type[ev.YearEvents]
        self.MonthEvents = type('MonthEvents', (ev.MonthEvents, ),
                                kw)  # type: Type[ev.MonthEvents]
        self.WeekEvents = type('WeekEvents', (ev.WeekEvents, ),
                               kw)  # type: Type[ev.WeekEvents]
        self.DayEvents = type('DayEvents', (ev.DayEvents, ),
                              kw)  # type: Type[ev.DayEvents]
        self.HourEvents = type('HourEvents', (ev.HourEvents, ),
                               kw)  # type: Type[ev.HourEvents]
        self.BitOpAnd = type('BitOpAnd', (ev.BitOpAnd, ),
                             kw)  # type: Type[ev.BitOpAnd]
        self.BitOpOr = type('BitOpOr', (ev.BitOpOr, ),
                            kw)  # type: Type[ev.BitOpOr]
        self.BitOpXor = type('BitOpXor', (ev.BitOpXor, ),
                             kw)  # type: Type[ev.BitOpXor]
        self.BitOpNot = type('BitOpNot', (ev.BitOpNot, ),
                             kw)  # type: Type[ev.BitOpNot]

    def mark_event(self,
                   event_name,
                   uuid,
                   timestamp=None,
                   track_hourly=None,
                   track_unique=None):
        """
        Marks an event as "happened" for a specific moment. The function
        stores the event for the day, week and month, and optionally
        for the hour, as well as the unique event.

        - event_name is the name of the event to track
        - uuid is the unique id of the subject (typically user id). The id
          should not be huge
        - timestamp is an optional moment of time which date should be used as
          a reference point, default is to `datetime.utcnow()`

        Examples:

            # Mark id 1 as active
            b.mark_event('active', 1)

            # Mark task completed for id 252
            b.mark_event('tasks:completed', 252)
        """
        self._mark(event_name, uuid, timestamp, 1, track_hourly, track_unique)

    def unmark_event(self,
                     event_name,
                     uuid,
                     timestamp=None,
                     track_hourly=None,
                     track_unique=None):
        """
        Marks an event as "not happened" for a specific moment. The function
        stores the event for the day, week and month, and optionally
        for the hour, as well as the unique event.
        """
        self._mark(event_name, uuid, timestamp, 0, track_hourly, track_unique)

    def _mark(self, event_name, uuid, timestamp, value, track_hourly,
              track_unique):
        if timestamp is None:
            timestamp = datetime.datetime.utcnow()
        if track_hourly is None:
            track_hourly = self.track_hourly
        if track_unique is None:
            track_unique = self.track_unique

        obj_classes = [self.MonthEvents, self.WeekEvents, self.DayEvents]
        if track_hourly:
            obj_classes.append(self.HourEvents)
        if track_unique:
            obj_classes.append(self.UniqueEvents)

        if self.pipe is None:
            pipe = self.connection.pipeline()
        else:
            pipe = self.pipe

        for obj_class in obj_classes:
            pipe.setbit(
                obj_class.from_date(event_name, timestamp).redis_key, uuid,
                value)

        if self.pipe is None:
            pipe.execute()

    def start_transaction(self):
        if self.pipe is not None:
            raise RuntimeError("Transaction already started")
        self.pipe = self.connection.pipeline()

    def commit_transaction(self):
        if self.pipe is None:
            raise RuntimeError("Transaction not started")
        self.pipe.execute()
        self.pipe = None

    def rollback_transaction(self):
        self.pipe = None

    @contextmanager
    def transaction(self):
        self.start_transaction()
        try:
            yield
            self.commit_transaction()
        except:
            self.rollback_transaction()
            raise

    def mark_unique(self, event_name, uuid):
        """
        Mark unique event as "happened with a user"

        Unique event (aka "user flag") is an event which doesn't depend on date.
        Can be used for storing user properties, A/B testing, extra filtering,
        etc.

        - event_name: The name of the event, could be "active" or "new_signups"
        - uuid:  a unique id, typically user id. The id should not be huge

        Example:

            # Mark id 42 as premium
            b.mark_unique('premium', 42)
        """
        self._mark_unique(event_name, uuid, value=1)

    def unmark_unique(self, event_name, uuid):
        """
        Mark unique event as "not happened with a user"

        Unique event (aka "user flag") is an event which doesn't depend on date.
        Can be used for storing user properties, A/B testing, extra filtering,
        etc.

        - event_name: The name of the event, could be "active" or "new_signups"
        - uuid:  a unique id, typically user id. The id should not be huge

        Example:

            # Mark id 42 as premium
            b.unmark_unique('premium', 42)
        """
        self._mark_unique(event_name, uuid, value=0)

    def _mark_unique(self, event_name, uuid, value):
        conn = self.connection if self.pipe is None else self.pipe
        redis_key = self.UniqueEvents(event_name).redis_key
        conn.setbit(redis_key, uuid, value)

    def get_event_names(self, prefix='', batch=10000):
        """
        Return the list of all event names, with no particular order. Optional
        `prefix` value is used to filter only subset of keys
        """
        expr = '{}{}*'.format(self.key_prefix, prefix)
        ret = set()
        for result in self.connection.scan_iter(match=expr, count=batch):
            result = result.decode()
            chunks = result.split('_')
            event_name = '_'.join(chunks[1:-1])
            if not event_name.startswith('bitop_'):
                ret.add(event_name)
        return sorted(ret)

    def delete_all_events(self):
        """
        Delete all events from the database.
        """
        keys = self.connection.keys('{}*'.format(self.key_prefix))
        if len(keys) > 0:
            self.connection.delete(*keys)

    def delete_temporary_bitop_keys(self):
        """
        Delete all temporary keys that are used when using bit operations.
        """
        keys = self.connection.keys('{}bitop_*'.format(self.key_prefix))
        if len(keys) > 0:
            self.connection.delete(*keys)

    def prefix_key(self, event_name, date):
        return '{}{}_{}'.format(self.key_prefix, event_name, date)
